docs/design/sql/14-automatic-query-resubmission.md

```diff
@@ -0,0 +1,157 @@
+# Automatic query resubmission
+
+|||
+|---|------------------- https://hazelcast.atlassian.net/browse/HZ-1098 https://hazelcast.atlassian.net/browse/HZ-1098 |
+|Related Github issues|https://github.com/hazelcast/hazelcast/pull/21635|
+|Document Status / Completeness| Complete                                          |
+|Requirement owner| Sandeep Akhouri                                              |
+|TDD Author|Viliam Ďurina|
+|Developer(s)| Krzysztof Ślusarski |
+|Quality Engineer| TODO                             |
+|Support Engineer| TODO                             |
+|Technical Reviewers| TODO                          |
+|Simulator or Soak Test PR(s) | TODO            |
+
+### Background
+#### Description
+
+A customer perceives that AP is NOT implemented correctly in Hazelcast. If the
+member hosting the primary partition is NOT available, their expectation is that
+data should be transparently served from the backup. When the cluster is
+changing topology, e.g., cluster member that contains the primary data is
+killed, the client throws errors.
+
+Automatic retries, with an upper bound on the number of retries, are implemented
+with many IMap operations on the client. However, with SQL, the client cannot do
+this, because it doesn't understand the query, and it doesn't know if the query
+is side-effect-free and it's safe to be retried. Secondly, the client might
+already have received some result rows, and if it retries the query, those rows
+will be received again.
+
+#### Scope
+
+The scope is to add support for automatic SQL retries in Java client. Other
+clients might follow, but it's not in the scope of the initial PR, as the
+requirement is currently only for Java client.
+
+There will be no retry mechanism for queries submitted through member instances.
+
+#### Terminology
+
+| Term           | Definition                            |
+|----------------|---------------------------------------|
+| Topology error | Query failure due to a member failure |
+
+### Technical Design
+
+The client doesn't understand the SQL statements it sends, and we want it to
+remain this way in order to simplify the implementation in multiple languages.
+We also don't want to retry statements by default, as not all are read-only.
+
+We will analyze three scenarios:
```

Comment:
What about the `SELECT` aggregations? Should there be a scenario for that too?

Reply:
Not sure what you mean here...

Reply:
I meant queries like `SELECT COUNT(*) ...`, are they handled the same with other `SELECT` queries?

Reply:
It's the same, from client's point of view, it's a query that starts with `SELECT`.